api/envoy/config/core/v3/socket_option.proto

```diff
@@ -74,6 +85,11 @@ message SocketOption {
   // The state in which the option will be applied. When used in BindConfig
   // STATE_PREBIND is currently the only valid value.
   SocketState state = 6 [(validate.rules).enum = {defined_only: true}];
+
+  // The socket type (see man 2 socket) to apply the socket option to. If
+  // this field is not specified, the option will be applied irrespective
+  // of the socket type.
```

Comment:
What happened when I use `DATAGRAM` socket option on listener with tcp address?

Reply:
In Envoy Mobile, we noticed some performance regression when setting the same send/receive buffer size for both UDP and TCP. So, this PR is to add a feature so that we can apply different send buffer/receive sizes to UDP and TCP.